deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AutoPromoter.java

```diff
@@ -122,85 +193,199 @@ boolean isCurrentDeployRetirable(String envId, DeployBean currDeployBean) throws
-            if (fromDeployId != null) {
-                DeployBean fromDeployBean = deployDAO.getById(fromDeployId);
-                if (fromDeployBean.getEnv_id().equals(predEnvBean.getEnv_id())) {
-                    currentDeployDate = fromDeployBean.getStart_date();
-                } else {
-                    LOG.info("Current deploy {} in env {} was not promoted from pred env {}, but from env {}! Use the current deploy startDate",
-                        currDeployBean.getDeploy_id(), currEnvBean.getEnv_id(), predEnvBean.getEnv_id(), fromDeployBean.getEnv_id());
-                    currentDeployDate = currDeployBean.getStart_date();
-                }
-            } else {
-                LOG.info("Current deploy {} in env {} was not promoted from anywhere! Use the current deploy startDate",
-                    currDeployBean.getDeploy_id(), currEnvBean.getEnv_id());
-                currentDeployDate = currDeployBean.getStart_date();
+        Date now = new Date();
+        long endTime = Long.MAX_VALUE;
+        String schedule = promoteBean.getSchedule();
+        if (!StringUtils.isEmpty(schedule)) {
+            //If we have a cron schedule set, looking into the due time of cron and only try to
+          // promote
+            //if we have a dued schedule.
+            long start_date = getScheduleStartTime(currEnvBean, currDeployBean, promoteBean);
+            //Get the time that we should deploy per schedule
+            Date autoDeployDueDate = getScheduledCheckDueTime(start_date, schedule);
+            if (autoDeployDueDate.getTime() != 0) {
+                endTime = autoDeployDueDate.getTime();
+            }
+            if (now.before(autoDeployDueDate)) {
+                return new PromoteResult()
+                    .withResultCode(PromoteResult.ResultCode.NotInScheduledTime);
             }
         }
 
-        DeployBean predDeployBean;
-        if (predPromoteBean.getDelay() != 0) {
-            long before = (System.currentTimeMillis() - predPromoteBean.getDelay() * 60 * 1000);
-            predDeployBean = getDeployCandidateDelayed(predEnvBean.getEnv_id(), before, currentDeployDate);
-            if (predDeployBean == null) {
-                LOG.debug("Could not find any deploy candidate within delay period for {}/{}", predEnvBean.getEnv_name(), predEnvBean.getStage_name());
-                return;
+        //Get the start time to find a deploy in preceded environment. If current deploy is promoted
+        //from preceded environment, use the last promoted deploy start_date in preceded
+      // environment.
+        //Otherwise (current deploy is not promoted from preceded environment), use the current
+      // deploy
+        // start_date
+        long
+            currentDeployDate =
+            getCurrentDeployStartDate(currDeployBean, precededEnvBean, currEnvBean);
+
+        DeployBean precededDeployBean;
+        if (promoteBean.getDelay() != 0) {
+            long
```

Comment:
I strongly think there is something wrong with format settings.  Following line is definitely more readable.

long before =
      Math.min(System.currentTimeMillis() - promoteBean.getDelay() * 60 * 1000, endTime);

It also follows google style which is adopted by Pinterest. http://google.github.io/styleguide/javaguide.html#s4.5-line-wrapping